python-sdk/src/astro/utils/table.py

```diff
@@ -1,97 +1,122 @@
+    op_kwargs: dict, python_callable: Callable, context: Context
+) -> BaseTable | None:
     """
-    When trying to "magically" determine the context of a decorator, we will try to find the first table.
-    This function attempts this by checking op_kwargs
+    Read op_kwargs and extract the tables.
 
     :param op_kwargs: user-defined operator's kwargs
-    :return: the first kwarg declared in the decorated method which is an
-        instance of BaseTable. Return None if non-existent
+    :param context: the context to use for resolving XComArgs
+    :return: first valid table found in op_kwargs.
     """
-    first_table = None
-    kwargs_of_table_type = [
-        op_kwargs[kwarg.name]
+    kwargs = [
+        op_kwargs[kwarg.name].resolve(context)
+        if isinstance(op_kwargs[kwarg.name], XComArg)
+        else op_kwargs[kwarg.name]
         for kwarg in inspect.signature(python_callable).parameters.values()
-        if isinstance(op_kwargs[kwarg.name], BaseTable)
     ]
-    if (
-        len(kwargs_of_table_type) == 1
-        or len({kwarg.conn_id for kwarg in kwargs_of_table_type}) == 1
-    ):
-        first_table = kwargs_of_table_type[0]
-    return first_table
+    tables = [kwarg for kwarg in kwargs if isinstance(kwarg, BaseTable)]
+
+    if _have_same_conn_id(tables):
+        return tables[0]
+    return None
 
 
-def _find_first_table_from_op_args(op_args: tuple) -> Optional[BaseTable]:
+def _find_first_table_from_parameters(
+    parameters: dict, context: Context
+) -> BaseTable | None:
     """
-    When trying to "magically" determine the context of a decorator, we will try to find the first table.
-    This function attempts this by checking op_args
+    Read parameters and extract the tables.
 
-    :param op_args: user-defined operator's args
-    :return: the first arg which is an instance of BaseTable. If there are no instances, return None.
+    :param parameters: a user-defined dictionary of parameters
+    :param context: the context to use for resolving XComArgs
+    :return: first valid table found in parameters.
     """
-    first_table = None
-    args_of_table_type = [arg for arg in op_args if isinstance(arg, BaseTable)]
-    # Check to see if all tables belong to same conn_id. Otherwise, we this can go wrong for cases
-    # 1. When we have tables from different DBs.
-    # 2. When we have tables from different conn_id, since they can be configured with different
-    # database/schema etc.
-    if (
-        len(args_of_table_type) == 1
-        or len({arg.conn_id for arg in args_of_table_type}) == 1
-    ):
-        first_table = args_of_table_type[0]
-    return first_table
+    params = [
+        param.resolve(context) if isinstance(param, XComArg) else param
+        for param in parameters.values()
+    ]
+    tables = [param for param in params if isinstance(param, BaseTable)]
+
+    if _have_same_conn_id(tables):
+        return tables[0]
+    return None
```

Comment:
It would be great if we could have tests covering the new scenarios introduced as part of this PR.

Reply:
This is not a new scenario. We had the same scenarios already before that, maybe you could not see it that well. But we were check for conn_id and either return the first table or none.

Reply:
Added tests for the xcom arg feature in find_first_table which you probably meant. See 504c9d1